services/src/main/java/org/jboss/windup/web/services/service/PackageService.java

```diff
@@ -0,0 +1,128 @@
+package org.jboss.windup.web.services.service;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.TreeMap;
+
+import javax.inject.Inject;
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+
+import org.jboss.forge.furnace.Furnace;
+import org.jboss.windup.web.addons.websupport.services.PackageDiscoveryService;
+import org.jboss.windup.web.furnaceserviceprovider.FromFurnace;
+import org.jboss.windup.web.services.model.ApplicationGroup;
+import org.jboss.windup.web.services.model.Package;
+import org.jboss.windup.web.services.model.PackageMetadata;
+import org.jboss.windup.web.services.model.RegisteredApplication;
+
+/**
+ * Service for manipulation with packages
+ *
+ * @author <a href="mailto:dklingenberg@gmail.com">David Klingenberg</a>
+ */
+public class PackageService
+{
+    @PersistenceContext
+    EntityManager entityManager;
+
+    @Inject
+    private Furnace furnace;
+
+    @Inject
+    @FromFurnace
+    private PackageDiscoveryService packageDiscoveryService;
+
+    /**
+     * Discovers packages in application
+     *
+     * @param application Application to discover packages in
+     * @return Collection of discovered packages
+     */
+    public Collection<Package> discoverPackages(RegisteredApplication application)
+    {
+        PackageMetadata metadata = application.getPackageMetadata();
+        metadata.setScanStatus(PackageMetadata.ScanStatus.IN_PROGRESS);
+        this.entityManager.merge(metadata);
+
+        ApplicationGroup appGroup = application.getApplicationGroup();
+        PackageMetadata appGroupMetadata = appGroup.getPackageMetadata();
+        appGroupMetadata.setScanStatus(PackageMetadata.ScanStatus.IN_PROGRESS);
+        this.entityManager.merge(appGroupMetadata);
+
+        String inputPath = application.getInputPath();
+        PackageDiscoveryService.PackageDiscoveryResult result = this.packageDiscoveryService.execute(inputPath);
+
+        Map<String, Package> packageMap = new TreeMap<>();
+
+        this.addPackagesToPackageMetadata(metadata, result.getKnownPackages(), packageMap);
+        this.addPackagesToPackageMetadata(metadata, result.getUnknownPackages(), packageMap);
+
+        this.addPackagesToPackageMetadata(appGroupMetadata, result.getKnownPackages(), packageMap);
+        this.addPackagesToPackageMetadata(appGroupMetadata, result.getUnknownPackages(), packageMap);
+
+        metadata.setScanStatus(PackageMetadata.ScanStatus.COMPLETE);
+        this.entityManager.merge(metadata);
+
+        PackageMetadata.ScanStatus finalStatus = appGroup.getApplications().stream()
+                    .map(app -> app.getPackageMetadata().getScanStatus())
+                    .reduce(PackageMetadata.ScanStatus.COMPLETE,
+                                (a, b) -> a == b ? PackageMetadata.ScanStatus.COMPLETE : PackageMetadata.ScanStatus.IN_PROGRESS);
```

Comment:
This accumulator seems a little off, if you have two IN_PROGRESS inputs they'll be equal and you'll end up with a value of COMPLETE. I think you want to check that a == b == COMPLETE.

Also, I think you can reach this point in the code with two apps in your app group such that one is COMPLETE and the other is QUEUED, and wind up with IN_PROGRESS for the app group. I wonder about the semantics of that (should it be QUEUED? should the accumulator really just be min(a,b)?) but I haven't thought about it too much.